---
title: "Vignette Title"
author:
- name: Letizia Girardi
- name: Andrea Policano
- name: Alberto Lupatin
output:
  BiocStyle::html_document
abstract: |
  
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Load Libraries

```{r}
library(biomaRt)
library(dplyr)
library(edgeR)
library(tidyr)
library(pathview)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(BSgenome.Hsapiens.UCSC.hg38)
library(biomaRt)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
library(MotifDb)
library(stringr)
library(Biostrings)
library(TFBSTools)
library(igraph)
library(clusterProfiler)
```


## TASK 1 - Loading RData file

The Liver_hepatocellular_carcinoma.RData file contains 3 dataframes: 
- *raw_count_df* with the raw RNA-Seq counts
- *c_anno_df* with the samples' names and condition
- *r_anno_df* with the ENSEMBL gene IDs, the length of the genes and the genes' symbols

```{r}
load("/Users/albertolupatin/Desktop/Bioinformatics Resources/Project/Liver_hepatocellular_carcinoma.RData")
```

## TASK 2 - Update raw_count_df and r_anno_df extracting only protein coding genes
In this first phase we filter and update gene annotations and raw counts data to focus on protein-coding genes using the BioMart database, facilitating subsequent analyses that require such specific gene information.
```{r}
# Select the BioMart database and the dataset to use
ensembl <- useMart(biomart="ensembl",dataset="hsapiens_gene_ensembl")

# Retrieve all the attributes needed for the project
query <- getBM(attributes = c("ensembl_gene_id",
                              "external_gene_name",
                              "gene_biotype",
                              "start_position",
                              "end_position",
                              "strand",
                              "description",
                              "version"),
               filters = "ensembl_gene_id",
               values = list(c(r_anno_df$ensembl_gene_id)),
               mart = ensembl)

head(query)
dim(query) # The query has around 62000 genes with the 8 attributes selected for each gene

# Select protein coding genes
query_pc <- query[which(query$gene_biotype=="protein_coding"),]

# Extract only protein coding genes
gene_names <- intersect(rownames(r_anno_df), query_pc$ensembl_gene_id)

# Update r_anno_fin with the new genes
r_anno_fin <- r_anno_df[gene_names, ]
dim(r_anno_fin) # The new dataframe has only 22000 thousand genes

# Update raw_counts_fin with the new genes
raw_counts_fin <- raw_counts_df[gene_names, ]
dim(raw_counts_fin)
```

## TASK 3 - Differential Expression Analysis

For this task we have to perform a differential expression analysis using edgeR package and select up- and
down-regulated genes using an adjusted p-value cutoff of 0.01, a log fold change ratio
>1.5 for up-regulated genes and < (-1.5) for down-regulated genes and a log CPM >1.

The results are visualizable through a volcano plot and annoted heatmap.

A new filtring of the raw data was performed: select a count threshold of counts grater than 20 and an occurrences of 5 sample in both groups

```{r}
# Filter raw counts data retaining only genes with raw count > 20...
count_thr <- 20
# ... in at least 5 Cases or 5 Control samples
repl_thr <- 5

filter_vec <- apply(raw_counts_fin,1,
    function(y) max(by(y, c_anno_df$condition, function(x) sum(x>=count_thr))))
# See statistics for the filtering
table(filter_vec) 

# Perform the filtering
filter_counts_df <- raw_counts_fin[filter_vec >= repl_thr,]
# Check the dimension of the filtered matrix 
dim(filter_counts_df)

# Apply the filter also on gene annotation
filter_anno_df <- r_anno_fin[rownames(filter_counts_df),]
dim(filter_anno_df)

# Create a DGEList object containing:
# - A numeric matrix with the read counts
# - A dataframe with samples' condition
# - A dataframe with samples
# - A dataframe giving annotation information for each gene

edge_c <- DGEList(counts = filter_counts_df,
                  group = c_anno_df$condition,
                  samples = c_anno_df,
                  genes = filter_anno_df)
edge_c

# Normalization with the edgeR package (TMM method)
edge_n <- calcNormFactors(edge_c,method="TMM")
edge_n

cpm_table <- as.data.frame(round(cpm(edge_n),2))
head(cpm_table)

filter_counts_df <- as.data.frame(filter_counts_df)
long_counts_df <- gather(filter_counts_df, key = "sample", value = "read_number")

# Define the experimental design matrix
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample

# Calculate dispersion and fit with edgeR for differential expression analysis
edge_d <- estimateDisp(edge_n,design)
edge_f <- glmQLFit(edge_d,design) 

# Definition of the contrast (conditions to be compared)
contro <- makeContrasts("case-control", levels=design) 

# Fit the model with generalized linear models
edge_t <- glmQLFTest(edge_f,contrast=contro)
DEGs <- as.data.frame(topTags(edge_t,n=20000)) # nolint: object_name_linter.
DEGs$class <- "="
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC>1.5)] = "+"
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC<(-1.5))] = "-"
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]


# MA Plot
xlabel <- "log2 avg CPM (A)"
ylabel <- "log2 FC"

par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(DEGs$logCPM, DEGs$logFC, xlab=xlabel, ylab=ylabel, 
     col=ifelse(DEGs$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="MA plot")
abline(h=0,lty=2,col="grey20")


cols <- c(rep("chartreuse4",50),rep("burlywood3",50))
pal <- c("blue","white","red") 
pal <- colorRampPalette(pal)(100)
heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%rownames(DEGs)[which(DEGs$class!="=")]),]),
        ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```
This heatmap shows a not optimal stratification of the samples


### Graphical representation of DEGs - Volcano Plot
The following volcano plot represents a high number of up-regulated genes, with a logFC > 1. Meanwhile, the number of down-regulated genes are visually lower.

```{r, fig.width = 10, fig.height = 6, fig.align = "center", echo = FALSE }
DEGs <- DEGs
xlabel <- "log2 FC control vs case"
ylabel <- "-log10 p-value"

par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(DEGs$logFC, -log(DEGs$PValue,base=10), xlab = xlabel, ylab = ylabel, 
     col = ifelse(DEGs$class=="=","grey85", ifelse(DEGs$class == "+", "indianred", "olivedrab4")), 
     pch = 20, frame.plot = TRUE, cex = 0.8, main = "Volcano plot")
abline(v=0,lty=2,col="grey20")
```

### Heatmap with DEG genes
The heatmap shows in the samples clustering a good stratification between control and test samples.
Notably, the there seem to be some differentially expressed genes (not colored in blue) in the test samples (yellow). 
Visually, a correlation can be found from some down-regulated genes and their sample type. Clearly some additional statistical test need to be performed to prove the correlation.

```{r, fig.width = 10, fig.height = 6, fig.align = "center", echo = FALSE }

cols <- ifelse(c_anno_df$condition == "control", "chartreuse4", "burlywood3") # Create a vector of colors based on control or test for annotation the heatmap
pal <- colorRampPalette(c("blue", "white", "red"))(100) # palette
heatmap(as.matrix(cpm_table[which(rownames(cpm_table) %in% rownames(DEGs)[which(DEGs$class != "=")]),]),
        ColSideColors = cols, cexCol = 0.5, margins = c(12, 4), col = pal, cexRow = 0.2)
```

## TASK 4 - Gene set Enrichment Analysis

# Filter genes depending on:
- Differential expression
- CLass
- Significant p-value

```{r, fig.width=10, fig.height=10, fig.align="center", echo = FALSE}
library(org.Hs.eg.db)

ensembl_gene_id <- rownames(DEGs)

convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id"),
                 filters=c("ensembl_gene_id"), 
                 values=rownames(DEGs),
                 mart = ensembl)

convert
DEGs <- merge(DEGs, convert, by = "ensembl_gene_id")
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),] # removing NAs
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),] # removing duplicates
DEGs

## Perform Gene Ontology enrichment analysis with Biological Process (BP)
##  upregulated genes
up_DEGs <- DEGs %>% filter(class == "+")
up_ego_BP <- enrichGO(gene = up_DEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

##  downregulated genes
down_DEGs <- DEGs %>% filter(class == "-")
down_ego_BP <- enrichGO(gene = down_DEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)

## Perform Gene Ontology enrichment analysis taking into account the Molecular Function (MF).
##  upregulated genes
up_ego_MF <- enrichGO(gene = up_DEGs$external_gene_name,
                      OrgDb = org.Hs.eg.db,
                      keyType = 'SYMBOL',
                      ont = "MF",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05)

##  downregulated genes
down_ego_MF <- enrichGO(gene = down_DEGs$external_gene_name,
                        OrgDb = org.Hs.eg.db,
                        keyType = 'SYMBOL',
                        ont = "MF",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05)


up_DEGs <- DEGs %>% filter(class == "+" & PValue <= 0.05)
down_DEGs <- DEGs %>% filter(class == "-" & PValue <= 0.05)

log_FC_pos <- up_DEGs$logFC

# Set the correct options for the pathviewer function
names(log_FC_pos) <- up_DEGs$entrezgene_id
# Check manually the image produced
pathview <- pathview(gene.data = log_FC_pos, 
         pathway.id = "hsa05202", 
         species = "human")
```

## TASK 5: Pathway Retrivial
The results obtained from the pathview algorithm show which are the biological component in which the up-regulated genes, 
obtained from step 4, are enriched in considering the case of Tubulin binding as "pathway id".
Further studies and researches may be useful for obtaining more insights regarding this event and its high correlation with Liver hepatocellular carcinoma
```{r}
#Starting with a filtering of the genes depending on the up or lower expression
log_FC_pos <- up_DEGs$logFC

#Set the correct options for the pathviewer function and then check manually the image produced
up_eKEGG <- enrichKEGG(gene = up_DEGs$ensembl_gene_id,
                organism = 'human',
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.1)
up_eKEGG # KEGG for finding the cell cycle pathway
first_pathway_id <- up_eKEGG@result$ID[1]
first_pathway_id
names(log_FC_pos) <- up_DEGs$entrezgene_id
pathview(gene.data = log_FC_pos, 
         pathway.id = "04210", 
         species = "human")
```

## TASK 6: Which TF have enriched score in the promoters of all up- and down-regulated genes

```{r, fig.width=10, fig.height=10, fig.align="center", echo = FALSE}
# Load the genome sequence and gene annotations
genome <- BSgenome.Hsapiens.UCSC.hg38
genes <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# Retrieve Gene Names
df <- getBM(attributes = c("external_gene_name",'entrezgene_id'),
            values=names(genes),filters ='entrezgene_id', mart = ensembl)
names(genes) <- df$external_gene_name[match(genes$gene_id,df$entrezgene_id)]

# Generate Promoter Regions
x <- promoters(genes,upstream = 500,downstream = 0)

# Extract the promoters' sequences for up- and down-regulated genes
up_reg <- subset(DEGs, class == "+")
dw_reg <- subset(DEGs, class == "-")

up_tf_names <- intersect(up_reg$external_gene_name, x@ranges@NAMES)
dw_tf_names <- intersect(dw_reg$external_gene_name, x@ranges@NAMES)

up_tf <- subset(up_reg, external_gene_name %in% up_tf_names)
dw_tf <- subset(dw_reg, external_gene_name %in% dw_tf_names)

up_promoters <- x[names(x) %in% up_tf_names]
dw_promoters <- x[names(x) %in% dw_tf_names]

up_promoter_seqs <- getSeq(genome, up_promoters)
dw_promoter_seqs <- getSeq(genome, dw_promoters)

# Perform motif enrichment analysis on the up- and down-regulated promoters
data(PWMLogn.hg19.MotifDb.Hsap)
up_enr <- motifEnrichment(up_promoter_seqs, PWMLogn.hg19.MotifDb.Hsap, score = "affinity")

# Generate a report of the enriched transcription factors and filter to keep only those with a p-value below 0.025
# We have selected only those above the 0.975 percentage of the data
report_up <- groupReport(up_enr, 1)
report_up <- report_up[order(report_up$raw.score, decreasing = TRUE),]
report_up <- report_up[report_up$p.value < 0.01,]

dw_enr <- motifEnrichment(dw_promoter_seqs, PWMLogn.hg19.MotifDb.Hsap, score = "affinity")
report_dw <- groupReport(dw_enr, 1)
report_dw <- report_dw[order(report_dw$raw.score, decreasing = TRUE),]
report_dw <- report_up[report_dw$p.value < 0.01,]

# Plot the first 5 promoters found
plot(report_up[1:10])
plot(report_dw[1:10])
```

The two graphs represent the 10 most representative motifs found for the up- and down-regulated genes.
In particular the graphs shows in order:
- The rank of the motif compared to the other motifs found
- The target TF
- The Positional Weight Matrix for the motif
- The Motif ID
- The raw enrichment score
- The p-value

These plots will be exploited to solve the next tasks.


## TASK 7: Compute the empirical distributions of scores
Selecting one among the top enriched TF for the up- and down-regulated, 
compute the empirical distributions of scores for all PWMs that we find in MotifDB 
for the selected TF and determining for all of them the distribution (log2) 
threshold cutoff at 99.75%.

```{r}
# UP-REGULATED genes; selecting PGAM2
tf_up_selected <- report_up@pwms$`Hsapiens-jolma2013-PGAM2`

# Extracting only the selected gene
mdb_human_up <- subset(MotifDb, organism=='Hsapiens' & geneSymbol=="PGAM2")
PWM_up <- toPWM(as.list(mdb_human_up))

# Applying the name of the selected genes to the dataset
names_up <- sapply(names(PWM_up), function(x) {
    parts <- str_split(x, "-")[[1]]
    return(parts[3])
})
names(PWM_up) <- names_up

# Compute the empirical distribution score for our set of motif
ecdf_up <- motifEcdf(PWM_up, organism = "hg19", quick = TRUE)
threshold_up <- log2(quantile(ecdf_up$`Hsapiens-jolma2013-PGAM2`, 0.9975))
scores_up <- motifScores(up_promoter_seqs, PWM_up, raw.scores = TRUE, cutoff = threshold_up)

```

The plot represent the first 5 TFs and the relative expression of the gene NFYA.

# TASK 8: Identify Up- and Down-regulated genes with a region in their promoter

```{r}
# Preparing the values for the following function 
matches <- c()
nomatches <- c()
best <- 0
best_scores <- 0

# Loop through each sequence in the up-regulated promoter sequences
for (n in 1:length(up_promoter_seqs)) {
  # Calculate motif scores for the current sequence using the given PWM and threshold
  scores <- motifScores(up_promoter_seqs[n], PWM_up, cutoff=threshold_up)
    # Check if any scores are above the threshold
    if (any(scores) > 0) {
     matches <- append(matches, n)
    if (sum(scores) > best) {
      best <- n
      best_scores <- motifScores(up_promoter_seqs[n], PWM_up, cutoff=threshold_up)
    }
  } else {
    nomatches <- append(nomatches, n)
  }
}

# Plot the motif scores for the best matching sequence
plotMotifScores(best_scores, legend.cex = 0.5)
```

Analogously, to the previous plot, this graph shows the same object but the function have been iterated for every promoter sequence retrieved from ensembl.

## 
```{r}
up_reg <- up_DEGs["external_gene_name"]
up_reg <- as.list(up_reg)

down_reg <- down_DEGs["external_gene_name"]
down_reg <- as.list(down_reg)

#Identify which are the genes across the dataframe which may be significant for us 
diff_express <- subset(DEGs, class == "+" & PValue <= 0.05 | class == "-" & PValue <= 0.05)

#Produce variables for their storing and manipulation
final_set <- diff_express["external_gene_name"]

final_list <- as.list(final_set)

#Produce a file in which the list of all the significant genes are present
reg <- write.table(final_list, file = "final_list.txt", sep = " ", row.names = F,
                   col.names = F)

```

After obtaining the file the procedure is: 
1) Go to STRING db;
2) Select the "Multiple proteins" option; 
3) Load the file and check the network


After exporting the tsv table of the coordinates and links from STRING and importing them in RStudio
```{r}
net_links <- read.delim("string_interactions_short.tsv")
head(net_links)


nodes <- read.delim("string_network_coordinates.tsv")
head(nodes)

# Build your network using the data from STRING
network <- graph_from_data_frame(d = net_links, vertices = nodes, directed = F)

#See the feature/characteristic of the network
class(network)
network

edge_attr(network)
vertex_attr(network)
graph_attr(network)

#Plot the network and compare the structure with the one obtained through STRING
plot(network, edge.width = 0.5, vertex.color = "cadetblue2", vertex.size = 5,
     vertex.frame.color = "darkgray", vertex.label = NA) 


#Plotting in this way will show you which are the external nodes's label
plot(network, vertex.shape = "none", vertex.label = final_set$external_gene_name, vertex.label.font = 1,
     vertex.label.cex = 0.5, vertex.label.color = "coral") 

###Now for the LCC
#Start by using RGB spectrum to assign a color to the future nodes
rgb_to_hex <- function(r, g, b) {
  rgb(r/255, g/255, b/255)
}

r <- 140
b <- 220
g <- 50
#Conversion to hexadecimal
color_hex <- rgb_to_hex(r, g, b)

#Identify the clusters and the Largest one
net_clusters <- clusters(network)
largest_cluster <- which.max(net_clusters$csize)
Larg_clust_nodes <- which(net_clusters$membership == largest_cluster)

#Produce a new network focused around the largest network
net_most_connected <- induced.subgraph(network, Larg_clust_nodes)
V(net_most_connected)$color <- rgb_to_hex(169, 239, 117)

#Find the node with the greatest degree, its label and assign to it a specific color 
nodes_degree <- degree(net_most_connected)
max_node <- which.max(nodes_degree)
max_node_ID <- V(net_most_connected)$label[max_node]

#V(net_most_connected)$label <- NA
#V(net_most_connected)$label[max_node] <- max_node_ID  --> tried to apply the label for 
#                                                          easier identification of the LCC 
#                                                          spoiler did not work, we searched it manually 
V(net_most_connected)$color[max_node] <- "red"

#Plot the new network with the wanted features and see where the LCC is located
plot(net_most_connected, main = "Largest Connected Component", 
     layout = layout_with_fr(net_most_connected),
     vertex.label = NA, 
     vertex.size = 2.8,
     vertex.color = V(net_most_connected)$color, 
     vertex.label.size = 1)

```

For the retrieval of the LCC label either try to implement it in the code(did not work, 
unknown causes) or go to the "net_most_connected" variable read the label associated to the node 
with max degree(being in this case the "int 36") and check the gene name, in our case it was 'ALB'  
     
                ||||||
                VVVVVV

The largest connected component is ALB, being connected with other 148 nodes.
ALB is known for being related to the production of the most abundant protein in
human blood: "Albumin", acting as a carrier for a wide range of endogenous molecule 
and helps in the regulation of plasma osmotic pressure. Characteristic of the HCC